### PR TITLE
Fix check of solc metadata tail

### PR DESCRIPTION
**Pull request by feliam**
Fixes 
https://github.com/trailofbits/manticore/issues/846
https://github.com/trailofbits/manticore/issues/844

Creates the need for
https://github.com/trailofbits/manticore/issues/847

---
This change is [Reviewable](https://reviewable.io/reviews/trailofbits/manticore/848)

### CODE REVIEW

File: manticore/ethereum.py [hunk 1]
```diff
@@ -607,6 +603,19 @@ def _compile(source_code, contract_name):
             :return: name, source_code, bytecode, srcmap, srcmap_runtime, hashes
         """
         solc = "solc"
+
+        #check solc version
+        supported_versions = ('0.4.18', '0.4.21')
+        installed_version = check_output(["solc", "--version"])
```
Comment (disconnect3d) [inline manticore/ethereum.py]: There is a variable called `solc` above which should be used here.

Anyway, as a long term solution we should probably have sth like `wrappers/solc.py` with e.g. `get_version`, `compile(source_code)` functions

**Comment by yan** [comment]
This touches on a lot of code that's being refactored in the py3 branch. Is this urgent? I'm thinking it might make sense to merge py3 first, then this.

**Comment by yan** [comment]
@feliam A part of this PR re-implements a refactoring i've done[1] in the py3 PR.  Would you rather me revert the py3 changes and go with your changes, or wait until py3 is merged and port to use those?

[1] https://github.com/trailofbits/manticore/pull/784/files#diff-6cf4568b5e8712514d226faa04e417fbR146

**Comment by yan** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[manticore/ethereum.py, line 144 at r2](https://beta.reviewable.io/reviews/trailofbits/manticore/848#-L9lok7v49fEgp8eLG1X:-L9lok7v49fEgp8eLG1Y:b-cldup8) ([raw file](https://github.com/trailofbits/manticore/blob/b22b4d7be65d677f66a79473aad734aa8a969720/manticore/ethereum.py#L144)):*
> ```Python
>     ''' Calculates what percentage of runtime_bytecode has been seen '''
>     end = None
>     if ''.join(runtime_bytecode[-43: -34]) == '\xa1\x65\x62\x7a\x7a\x72\x30\x58\x20' \
> ```

Use the new `_remove_tail` here as well?

---

*[manticore/ethereum.py, line 170 at r2](https://beta.reviewable.io/reviews/trailofbits/manticore/848#-L9lnoo65IKTLYxdKnWa:-L9lnoo65IKTLYxdKnWb:bhp0nyb) ([raw file](https://github.com/trailofbits/manticore/blob/b22b4d7be65d677f66a79473aad734aa8a969720/manticore/ethereum.py#L170)):*
> ```Python
>         self.srcmap = self.__build_source_map(self.init_bytecode, srcmap)
> 
>     def _remove_tail(self, bytecode):
> ```

What is in the tail? Metadata? Perhaps rename this to `without_metadata` or another descriptive name. (I assumed it was metadata and it's what i named a similar method in the py3 branch)

---

*[manticore/ethereum.py, line 614 at r2](https://beta.reviewable.io/reviews/trailofbits/manticore/848#-L9loTlKA0Q9_xN317KX:-L9loTlKA0Q9_xN317KY:bqfwc11) ([raw file](https://github.com/trailofbits/manticore/blob/b22b4d7be65d677f66a79473aad734aa8a969720/manticore/ethereum.py#L614)):*
> ```Python
>         installed_version = check_output(["solc", "--version"])
>         for line in installed_version.split('\n'):
>             if 'ersion:' in line:
> ```

Just check for `:` being in `line`? Has other `solc` versions used a lowercase `Version:` prefix?

---

*[manticore/ethereum.py, line 617 at r2](https://beta.reviewable.io/reviews/trailofbits/manticore/848#-L9lodSWAsU-h0acbxek:-L9lodSWAsU-h0acbxel:b9akk88) ([raw file](https://github.com/trailofbits/manticore/blob/b22b4d7be65d677f66a79473aad734aa8a969720/manticore/ethereum.py#L617)):*
> ```Python
>                 installed_version = line.split(":")[1].split('+')[0]
>                 break
>         if installed_version not in supported_versions:
> ```

If the version isn't parsed correctly above, this will always fail.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/trailofbits/manticore/848)*

**Comment by feliam** [comment]
*[manticore/ethereum.py, line 614 at r2](https://beta.reviewable.io/reviews/trailofbits/manticore/848#-L9loTlKA0Q9_xN317KX:-L9mCtpe1NbCbNsRE_o7:b-l10zsx) ([raw file](https://github.com/trailofbits/manticore/blob/b22b4d7be65d677f66a79473aad734aa8a969720/manticore/ethereum.py#L614)):*
<details><summary><i>Previously, yan (Yan Ivnitskiy) wrote…</i></summary><blockquote>

Just check for `:` being in `line`? Has other `solc` versions used a lowercase `Version:` prefix?
</blockquote></details>

Not sure about version vs. Version.  This is what I've checked:
Version: 0.4.21+commit.dfe3193c.Linux.g++
Version: 0.4.20+commit.3155dd80.Linux.g++
Version: 0.4.19+commit.c4cbbb05.Linux.g++
Version: 0.4.18+commit.9cf6e910.Linux.g++
Version: 0.4.16+commit.d7661dd9.Linux.g++
Version: 0.4.15+commit.8b45bddb.Linux.g++
Version: 0.4.14+commit.c2215d46.Linux.g++
Version: 0.4.13+commit.0fb4cb1a.Linux.g++
Version: 0.4.12+commit.194ff033.Linux.g++
Version: 0.4.11+commit.68ef5810.Linux.g++
Version: 0.4.10+commit.9e8cc01b.Linux.g++
I did not include the "V" trying to be conservative.  In fact I should do line.upper()!

Normally it prints another line, something like:
"solc, the solidity compiler commandline interface"
I'm not sure if in the Windows version the extra lines have a ":" 

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/trailofbits/manticore/848)*

**Comment by feliam** [comment]
*[manticore/ethereum.py, line 617 at r2](https://beta.reviewable.io/reviews/trailofbits/manticore/848#-L9lodSWAsU-h0acbxek:-L9mJAYd9pdY4jmRet4B:b-co0vsz) ([raw file](https://github.com/trailofbits/manticore/blob/b22b4d7be65d677f66a79473aad734aa8a969720/manticore/ethereum.py#L617)):*
<details><summary><i>Previously, yan (Yan Ivnitskiy) wrote…</i></summary><blockquote>

If the version isn't parsed correctly above, this will always fail.
</blockquote></details>

So that solc is not supported ?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/trailofbits/manticore/848)*

**Comment by feliam** [comment]
*[manticore/ethereum.py, line 617 at r2](https://beta.reviewable.io/reviews/trailofbits/manticore/848#-L9lodSWAsU-h0acbxek:-L9uPeAD16G2KVAxO6Ug:bv1q3xg) ([raw file](https://github.com/trailofbits/manticore/blob/b22b4d7be65d677f66a79473aad734aa8a969720/manticore/ethereum.py#L617)):*
<details><summary><i>Previously, feliam (feliam) wrote…</i></summary><blockquote>

So that solc is not supported ?
</blockquote></details>

Done. Added a regular expression to parse the version so if it is not supported log (not currently as log does not work) that it is unsupported

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/trailofbits/manticore/848)*

**Comment by feliam** [comment]
Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions.

---

*[manticore/ethereum.py, line 609 at r1](https://beta.reviewable.io/reviews/trailofbits/manticore/848#-L8YWc8zUxKHQ22kkj0J-r1-609:-L9uPboZEqbOmWn05Vqr:b-896fix) ([raw file](https://github.com/trailofbits/manticore/blob/dc7ad49974dd7448de3f55644ef1aa15169e6a5b/manticore/ethereum.py#L609)):*
<details><summary><i>Previously, disconnect3d (Disconnect3d) wrote…</i></summary><blockquote>

There is a variable called `solc` above which should be used here.

Anyway, as a long term solution we should probably move this logic to sth like `wrappers/solc.py` with e.g. `get_version`, `compile(source_code)` functions
</blockquote></details>

Done.

---

*[manticore/ethereum.py, line 144 at r2](https://beta.reviewable.io/reviews/trailofbits/manticore/848#-L9lok7v49fEgp8eLG1X:-L9uPZiJ82CyryyKNGXU:b-896fix) ([raw file](https://github.com/trailofbits/manticore/blob/b22b4d7be65d677f66a79473aad734aa8a969720/manticore/ethereum.py#L144)):*
<details><summary><i>Previously, yan (Yan Ivnitskiy) wrote…</i></summary><blockquote>

Use the new `_remove_tail` here as well?
</blockquote></details>

Done.

---

*[manticore/ethereum.py, line 170 at r2](https://beta.reviewable.io/reviews/trailofbits/manticore/848#-L9lnoo65IKTLYxdKnWa:-L9uPaAy4nwCDVPOHzuE:b-896fix) ([raw file](https://github.com/trailofbits/manticore/blob/b22b4d7be65d677f66a79473aad734aa8a969720/manticore/ethereum.py#L170)):*
<details><summary><i>Previously, yan (Yan Ivnitskiy) wrote…</i></summary><blockquote>

What is in the tail? Metadata? Perhaps rename this to `without_metadata` or another descriptive name. (I assumed it was metadata and it's what i named a similar method in the py3 branch)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/trailofbits/manticore/848)*

**Comment by feliam** [comment]
*[manticore/ethereum.py, line 609 at r1](https://beta.reviewable.io/reviews/trailofbits/manticore/848#-L8YWc8zUxKHQ22kkj0J-r1-609:-LAK_Vwr0sfGkhcgZdTW:b32w0x) ([raw file](https://github.com/trailofbits/manticore/blob/dc7ad49974dd7448de3f55644ef1aa15169e6a5b/manticore/ethereum.py#L609)):*
<details><summary><i>Previously, feliam (feliam) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

OK

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/trailofbits/manticore/848)*

**Pull request merged**